crates/ruff_python_parser/src/parser/recovery.rs

```diff
@@ -0,0 +1,165 @@
+                keys.push(None);
+                values.push(Expr::Name(ast::ExprName {
+                    range: rest.range,
+                    id: rest.id,
+                    ctx: ExprContext::Store,
+                }));
+            }
+            Expr::Dict(ast::ExprDict {
+                range,
+                keys,
+                values,
+            })
+        }
+        Pattern::MatchClass(ast::PatternMatchClass {
+            range,
+            cls,
+            arguments,
+        }) => Expr::Call(ast::ExprCall {
+            range,
+            func: cls,
+            arguments: ast::Arguments {
+                range: arguments.range,
+                args: arguments
+                    .patterns
+                    .into_iter()
+                    .map(pattern_to_expr)
+                    .collect(),
+                keywords: arguments
+                    .keywords
+                    .into_iter()
+                    .map(|keyword_pattern| ast::Keyword {
+                        range: keyword_pattern.range,
+                        arg: Some(keyword_pattern.attr),
+                        value: pattern_to_expr(keyword_pattern.pattern),
+                    })
+                    .collect(),
+            },
+        }),
+        Pattern::MatchStar(ast::PatternMatchStar { range, name }) => {
+            if let Some(name) = name {
+                Expr::Starred(ast::ExprStarred {
+                    range,
+                    value: Box::new(Expr::Name(ast::ExprName {
+                        range: name.range,
+                        id: name.id,
+                        ctx: ExprContext::Store,
+                    })),
+                    ctx: ExprContext::Store,
+                })
+            } else {
+                Expr::Starred(ast::ExprStarred {
+                    range,
+                    value: Box::new(Expr::Name(ast::ExprName {
+                        range: TextRange::new(range.end() - "_".text_len(), range.end()),
+                        id: "_".to_string(),
+                        ctx: ExprContext::Store,
+                    })),
+                    ctx: ExprContext::Store,
+                })
+            }
+        }
+        Pattern::MatchAs(ast::PatternMatchAs {
+            range,
+            pattern,
+            name,
+        }) => {
+            if let Some(pattern) = pattern {
+                pattern_to_expr(*pattern)
+            } else if let Some(name) = name {
+                Expr::Name(ast::ExprName {
+                    range: name.range,
+                    id: name.id,
+                    ctx: ExprContext::Store,
+                })
+            } else {
+                Expr::Name(ast::ExprName {
+                    range,
+                    id: "_".to_string(),
+                    ctx: ExprContext::Store,
+                })
+            }
+        }
+        Pattern::MatchOr(ast::PatternMatchOr { patterns, .. }) => {
```

Comment:
Does that mean we drop the `as <name>` part in this case? I'm not quiet sure what we should do in this case to still get as nice error recovery. What are the cases where we need this branch? What I understand is that it is for:

```python
match invalid_rhs_pattern:
    case {"x" as y:1 + Foo()}:
        pass
```

I think it might be best to disallow `As` when parsing the `key` and make recovery panic (or return a Result) if you try to convert a `MatchAs`.

Reply:
Fixed in https://github.com/astral-sh/ruff/pull/10523